subprojects/frameworklauncher/yarn/src/main/java/com/microsoft/frameworklauncher/common/exit/AMDiagnostics.java

```diff
@@ -0,0 +1,220 @@
+  }
+
+  public void setApplicationExitTriggerMessage(String applicationExitTriggerMessage) {
+    this.applicationExitTriggerMessage = applicationExitTriggerMessage;
+  }
+
+  public String getApplicationExitTriggerTaskRoleName() {
+    return applicationExitTriggerTaskRoleName;
+  }
+
+  public void setApplicationExitTriggerTaskRoleName(String applicationExitTriggerTaskRoleName) {
+    this.applicationExitTriggerTaskRoleName = applicationExitTriggerTaskRoleName;
+  }
+
+  public Integer getApplicationExitTriggerTaskIndex() {
+    return applicationExitTriggerTaskIndex;
+  }
+
+  public void setApplicationExitTriggerTaskIndex(Integer applicationExitTriggerTaskIndex) {
+    this.applicationExitTriggerTaskIndex = applicationExitTriggerTaskIndex;
+  }
+
+  public static void limitSerializationMaxBytes(Integer serializationMaxBytes) {
+    AMDiagnostics.serializationMaxBytes = serializationMaxBytes;
+  }
+
+  public static String generateAndSerialize(
+      int applicationExitCode,
+      String applicationExitDiagnostics,
+      String applicationExitTriggerMessage,
+      String applicationExitTriggerTaskRoleName,
+      Integer applicationExitTriggerTaskIndex) {
+    String serializedObj = LANDMARK_START + WebCommon.toJson(generate(
+        applicationExitCode,
+        applicationExitDiagnostics,
+        applicationExitTriggerMessage,
+        applicationExitTriggerTaskRoleName,
+        applicationExitTriggerTaskIndex)) + LANDMARK_END;
+
+    if (serializationMaxBytes != null &&
+        serializationMaxBytes < serializedObj.length()) {
+      // Best Effort to truncate
+      LOGGER.logWarning(
+          "Raw AMDiagnostics bytes %s is large than max bytes %s, " +
+              "applicationExitDiagnostics will be truncated",
+          serializedObj.length(), serializationMaxBytes);
+
+      Integer exceededBytes = serializedObj.length() - serializationMaxBytes;
+      Integer retainBytes = applicationExitDiagnostics.length() - exceededBytes;
+      if (retainBytes >= 0) {
+        // Truncate applicationExitDiagnostics is enough
+        applicationExitDiagnostics = applicationExitDiagnostics.substring(0, retainBytes);
+      } else {
+        applicationExitDiagnostics = "";
+        LOGGER.logWarning(
+            "Truncated AMDiagnostics bytes %s is still large than max bytes %s, " +
+                "applicationExitTriggerMessage will be truncated",
+            serializationMaxBytes - retainBytes, serializationMaxBytes);
+
+        retainBytes = applicationExitTriggerMessage.length() + retainBytes;
+        if (retainBytes >= 0) {
+          // Truncate applicationExitDiagnostics and applicationExitTriggerMessage is enough
+          applicationExitTriggerMessage = applicationExitTriggerMessage.substring(0, retainBytes);
+        } else {
+          applicationExitTriggerMessage = "";
+
+          LOGGER.logError(
+              "Truncated AMDiagnostics bytes %s is still large than max bytes %s, " +
```

Comment:
Just curious about how RM will truncated the RM. After the message is truncated, it will still be the valid Json?

Reply:
May not be valid Json.

Reply:
Anyway this should be avoided by properly setting YARN and Launcher config.